File: bhashalens_app/lib/services/circuit_breaker.dart
import 'dart:async';
import 'package:flutter/foundation.dart';

/// Circuit breaker states
enum CircuitState {
  closed,  // Normal operation, requests pass through
  open,    // Circuit is open, requests fail fast
  halfOpen // Testing if service has recovered
}

/// Circuit breaker to prevent cascading failures
/// Implements the circuit breaker pattern for cloud service calls
class CircuitBreaker {
  final String name;
  final int failureThreshold;
  final Duration timeout;
  final Duration resetTimeout;
  
  CircuitState _state = CircuitState.closed;
  int _failureCount = 0;
  int _successCount = 0;
  DateTime? _lastFailureTime;
  Timer? _resetTimer;

  CircuitBreaker({
    required this.name,
    this.failureThreshold = 5,
    this.timeout = const Duration(seconds: 5),
    this.resetTimeout = const Duration(seconds: 30),
  });

  /// Get current circuit state
  CircuitState get state => _state;

  /// Get failure count
  int get failureCount => _failureCount;

  /// Check if circuit is open
  bool get isOpen => _state == CircuitState.open;

  /// Execute an operation through the circuit breaker
  Future<T> execute<T>(Future<T> Function() operation) async {
    // Check if circuit should transition from open to half-open
    if (_state == CircuitState.open) {
      if (_shouldAttemptReset()) {
        _transitionToHalfOpen();
      } else {
        throw CircuitBreakerOpenException(
          'Circuit breaker "$name" is open',
          failureCount: _failureCount,
        );
      }
    }

    try {
      final result = await operation().timeout(timeout);
      _onSuccess();
      return result;
    } catch (e) {
      _onFailure();
      rethrow;
    }
  }

  /// Handle successful operation
  void _onSuccess() {
    _failureCount = 0;
    
    if (_state == CircuitState.halfOpen) {
      _successCount++;
      // After 2 successful calls in half-open, close the circuit
      if (_successCount >= 2) {
        _transitionToClosed();
      }
    }
  }

  /// Handle failed operation
  void _onFailure() {
    _failureCount++;
    _lastFailureTime = DateTime.now();
    _successCount = 0;

    if (_state == CircuitState.halfOpen) {
      // Failure in half-open state immediately opens circuit
      _transitionToOpen();
    } else if (_failureCount >= failureThreshold) {
      _transitionToOpen();
    }
  }

  /// Check if enough time has passed to attempt reset
  bool _shouldAttemptReset() {
    if (_lastFailureTime == null) return false;
    
    final timeSinceLastFailure = DateTime.now().difference(_lastFailureTime!);
    return timeSinceLastFailure >= resetTimeout;
  }

  /// Transition to closed state
  void _transitionToClosed() {
    debugPrint('Circuit breaker "$name": CLOSED');
    _state = CircuitState.closed;
    _failureCount = 0;
    _successCount = 0;
    _resetTimer?.cancel();
  }

  /// Transition to open state
  void _transitionToOpen() {
    debugPrint('Circuit breaker "$name": OPEN (failures: $_failureCount)');
    _state = CircuitState.open;
    
    // Schedule automatic transition to half-open
    _resetTimer?.cancel();
    _resetTimer = Timer(resetTimeout, () {
      if (_state == CircuitState.open) {
        _transitionToHalfOpen();
      }
    });
  }

  /// Transition to half-open state
  void _transitionToHalfOpen() {
    debugPrint('Circuit breaker "$name": HALF-OPEN (testing recovery)');
    _state = CircuitState.halfOpen;
    _successCount = 0;
  }

  /// Manually reset the circuit breaker
  void reset() {
    _transitionToClosed();
  }

  /// Dispose resources
  void dispose() {
    _resetTimer?.cancel();
  }
}

/// Exception thrown when circuit breaker is open
class CircuitBreakerOpenException implements Exception {
  final String message;
  final int failureCount;

  CircuitBreakerOpenException(this.message, {required this.failureCount});

  @override
  String toString() => 'CircuitBreakerOpenException: $message';
}

/// Circuit breaker registry to manage multiple circuit breakers
class CircuitBreakerRegistry {
  static final CircuitBreakerRegistry _instance = 
      CircuitBreakerRegistry._internal();
  
  factory CircuitBreakerRegistry() => _instance;
  
  CircuitBreakerRegistry._internal();

  final Map<String, CircuitBreaker> _breakers = {};

  /// Get or create a circuit breaker
  CircuitBreaker getOrCreate(
    String name, {
    int failureThreshold = 5,
    Duration timeout = const Duration(seconds: 5),
    Duration resetTimeout = const Duration(seconds: 30),
  }) {
    return _breakers.putIfAbsent(
      name,
      () => CircuitBreaker(
        name: name,
        failureThreshold: failureThreshold,
        timeout: timeout,
        resetTimeout: resetTimeout,
      ),
    );
  }

  /// Get circuit breaker by name
  CircuitBreaker? get(String name) => _breakers[name];

  /// Reset all circuit breakers
  void resetAll() {
    for (final breaker in _breakers.values) {
      breaker.reset();
    }
  }

  /// Get status of all circuit breakers
  Map<String, CircuitState> getStatus() {
    return Map.fromEntries(
      _breakers.entries.map((e) => MapEntry(e.key, e.value.state)),
    );
  }

  /// Dispose all circuit breakers
  void dispose() {
    for (final breaker in _breakers.values) {
      breaker.dispose();
    }
    _breakers.clear();
  }
}

File: infrastructure/lambda/layers/shared/python/shared/circuit_breaker.py
"""
Shared Circuit Breaker utility
"""
import time
import logging

logger = logging.getLogger(__name__)

from enum import Enum
import threading

class CircuitState(Enum):
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"

class CircuitBreaker:
    def __init__(self, failure_threshold=3, recovery_timeout=60):
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.failure_count = 0
        self.last_failure_time = 0
        self.state = CircuitState.CLOSED
        self._lock = threading.Lock()
        self._half_open_test_in_progress = False
    def record_success(self):
        """Reset the breaker on success"""
        with self._lock:
            self.failure_count = 0
            self.state = "CLOSED"
            self._half_open_test_in_progress = False

    def record_failure(self):
        """Record a failure and potentially open the breaker"""
        with self._lock:
            if self.state == "HALF_OPEN":
                self.state = "OPEN"
                self.last_failure_time = time.time()
                self._half_open_test_in_progress = False
                logger.warning("Circuit Breaker re-OPENED after HALF_OPEN test failure")
                return
            
            self.failure_count += 1
            self.last_failure_time = time.time()
            
            if self.failure_count >= self.failure_threshold:
                self.state = "OPEN"
                logger.warning(f"Circuit Breaker OPENED after {self.failure_count} failures")
    def is_allowed(self) -> bool:
        """Check if calls are currently allowed"""
        with self._lock:
            if self.state == "CLOSED" or self.state == CircuitState.CLOSED:
                return True
                
            if self.state == CircuitState.OPEN:
                # Check if enough time has passed to transition to HALF_OPEN
                if time.time() - self.last_failure_time > self.recovery_timeout:
                    self.state = CircuitState.HALF_OPEN
                    self._half_open_test_in_progress = True
                    logger.info("Circuit Breaker transitioned to HALF_OPEN")
                    return True
                return False
                
            if self.state == CircuitState.HALF_OPEN:
                if not self._half_open_test_in_progress:
                    self._half_open_test_in_progress = True
                    return True
                return False
                
            return False
